pkg/controllers/cloud_config_sync_controller.go

```diff
@@ -60,39 +62,80 @@ func (r *CloudConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request)
-		if err := r.Get(ctx, openshiftUnmanagedCMKey, sourceCM); err != nil {
-			klog.Errorf("unable to get cloud-config for sync")
+	sourceCM := &corev1.ConfigMap{}
+	managedConfigFound := false
+
+	// NOTE: We know that there is some transformation logic in place in the
+	// Cluster Config Operator (CCO) for AWS and Azure. We have not implemented
+	// this logic here yet so we've intentionally chosen to lookup up config
+	// from the (CCO-) managed namespace **only for these cloud platforms**
+	// TODO: Drop this once we implement the AWS and Azure transformers here in
+	// CCCMO, allowing us to drop this kinda-sorta reliance on CCO stuff. We
+	// may also wish to merge the use of cloudConfigTransformerFn into the
+	// prepareSourceConfigMap helper function
+	if cloudConfigTransformerFn == nil {
+		defaultSourceCMObjectKey := client.ObjectKey{
+			Name:      managedCloudConfigMapName,
+			Namespace: OpenshiftManagedConfigNamespace,
+		}
+		if err := r.Get(ctx, defaultSourceCMObjectKey, sourceCM); err == nil {
+			managedConfigFound = true
+		} else if errors.IsNotFound(err) {
+			klog.Warningf("managed cloud-config is not found, falling back to infrastructure config")
+		} else if err != nil {
+			klog.Errorf("unable to get managed cloud-config for sync")
 			if err := r.setDegradedCondition(ctx); err != nil {
 				return ctrl.Result{}, fmt.Errorf("failed to set conditions for cloud config controller: %v", err)
 			}
 			return ctrl.Result{}, err
 		}
-		sourceCM, err = r.prepareSourceConfigMap(sourceCM, infra)
-		if err != nil {
+	}
+
+	if !managedConfigFound {
+		openshiftUnmanagedCMKey := client.ObjectKey{
+			Name:      infra.Spec.CloudConfig.Name,
+			Namespace: OpenshiftConfigNamespace,
+		}
+		if err := r.Get(ctx, openshiftUnmanagedCMKey, sourceCM); err != nil {
+			klog.Errorf("unable to get cloud-config for sync")
 			if err := r.setDegradedCondition(ctx); err != nil {
 				return ctrl.Result{}, fmt.Errorf("failed to set conditions for cloud config controller: %v", err)
 			}
 			return ctrl.Result{}, err
 		}
-	} else if err != nil {
-		klog.Errorf("unable to get managed cloud-config for sync")
+	}
+
+	sourceCM, err = r.prepareSourceConfigMap(sourceCM, infra)
+	if err != nil {
 		if err := r.setDegradedCondition(ctx); err != nil {
 			return ctrl.Result{}, fmt.Errorf("failed to set conditions for cloud config controller: %v", err)
 		}
 		return ctrl.Result{}, err
 	}
 
+	if cloudConfigTransformerFn != nil {
+		// We ignore stuff in sourceCM.BinaryData. This isn't allowed to
+		// contain any key that overlaps with those found in sourceCM.Data and
+		// we're not expecting users to put their data in the former.
+		output, err := cloudConfigTransformerFn(sourceCM.Data[defaultConfigKey], infra)
+		if err != nil {
+			if err := r.setDegradedCondition(ctx); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to set conditions for cloud config controller: %v", err)
+			}
+			return ctrl.Result{}, err
```

Comment:
I think we have a critical problem here: this value of `err` contains information which is required by the user, but we're not giving it to them. I believe controller-runtime will log this value, but logs are not user-visible. We need to somehow relate this error message to the condition and/or generate an event containing it.

I think we can merge this PR without fixing this, but I don't think we can GA without fixing it.